test/run_test.py

```diff
@@ -1435,79 +1437,105 @@ def main():
+            if test_failed and not options.continue_through_error:
+                pool.terminate()
+
+        try:
+            os.environ["NUM_PARALLEL_PROCS"] = str(NUM_PROCS)
+            for test in selected_tests_parallel:
+                options_clone = copy.deepcopy(options)
+                if can_run_in_pytest(test):
+                    options_clone.pytest = True
+                pool.apply_async(
+                    run_test_module,
+                    args=(test, test_directory, options_clone),
+                    callback=parallel_test_completion_callback,
+                )
+            pool.close()
+            pool.join()
+            del os.environ["NUM_PARALLEL_PROCS"]
+
+            if not options.continue_through_error and len(failure_messages) != 0:
+                raise RuntimeError(
+                    "\n".join(failure_messages)
+                    + "\n\nTip: You can keep running tests even on failure by "
+                    "passing --keep-going to run_test.py.\n"
+                    "If running on CI, add the 'keep-going' label to "
+                    "your PR and rerun your jobs."
+                )
+
+            for test in selected_tests_serial:
+                options_clone = copy.deepcopy(options)
+                if can_run_in_pytest(test):
+                    options_clone.pytest = True
+                err_message = run_test_module(test, test_directory, options_clone)
+                test_failed = handle_error_messages(err_message)
+                if test_failed and not options.continue_through_error:
+                    raise RuntimeError(err_message)
+
+        finally:
             pool.terminate()
+            pool.join()
 
+        return failure_messages
+
+    # First run the prioritized tests, then the remaining tests.
     try:
-        os.environ["NUM_PARALLEL_PROCS"] = str(NUM_PROCS)
-        for test in selected_tests_parallel:
-            options_clone = copy.deepcopy(options)
-            if can_run_in_pytest(test):
-                options_clone.pytest = True
-            pool.apply_async(
-                run_test_module,
-                args=(test, test_directory, options_clone),
-                callback=parallel_test_completion_callback,
-            )
-        pool.close()
-        pool.join()
-        del os.environ["NUM_PARALLEL_PROCS"]
+        failure_messages = run_tests(
+            prioritized_tests, test_directory, options, "Prioritized tests"
+        )
 
-        if not options.continue_through_error and len(failure_messages) != 0:
-            raise RuntimeError(
-                "\n".join(failure_messages)
-                + "\n\nTip: You can keep running tests even on failure by "
-                "passing --keep-going to run_test.py.\n"
-                "If running on CI, add the 'keep-going' label to "
-                "your PR and rerun your jobs."
+        if options.continue_through_error or len(failure_messages) == 0:
```

Comment:
I think this check is redundant as a `RuntimeError` will be raised and the test ends anyway when `options.continue_through_error` is false

Reply:
True. Removed